### DISCUSSION TITLE
Question on `MAP` posterior approximator

### DISCUSSION THREAD

**Discussion by PaulScemama** [awslabs/fortuna#77] [Q&A] [answered]
Hi! I trained a MAP posterior approximator and was looking at predictive samples:

```
map_model.prective.sample(data_loader.to_inputs_loader())
```
And I observed that each time I call this I get different answers. Where is the randomness coming from if a MAP approximation only sees the parameters $\mathbf{\theta}$ point-estimates? I looked at the documentation but couldn't find anything. 

Thanks in advance!

**Comment by gianlucadetommaso** [comment, marked as answer]
Hi Paul, you are sampling target variables $y^{(i)}$ from the predictive distribution $p(y|x, D)$, where $x$ is a certain input for which you want to get target samples, and $D$ is your training data. In order to sample from the predictive distribution, you can actually sample from the joint distribution $p(y, \theta|x, D) = p(y|\theta, x)p(\theta|D)$, where $\theta$ are model parameters. This means that, in order to obtain a sample $y^{(i)}$ given $x$ and $D$, you can first sample $\theta^{(j)}$ from the posterior $p(\theta|D)$, then sample $y^{i}$ from the likelihood $p(y|\theta^{(j)}, x)$.

As you correctly said, there is no randomness in sampling from the MAP approximation of the posterior, since you always get the sample parameters $\theta^{MAP}$. However, there is still randomness when sampling from the likelihood $p(y|\theta^{MAP}, x)$.

I hope this explains it. Please let me know if it is clear.

**Reply by PaulScemama** [reply]
Thank you very much for the clear explanation! This makes sense, thank you :). One follow-up question: is the randomness when sampling from the likelihood from a fixed variance that's calculated (i.e. $\beta$?) In other words, is it assuming fixed variance due to (e.g.) label noise?

**Reply by gianlucadetommaso** [reply]
From your question, I assume you are doing regression? In such case, the likelihood variance model is also learned (unless you don't want to, in which case you can just pass a hyperparameter). In `ProbRegressor` you need to pass both `model`, which is the mean likelihood model, and `likelihood_log_variance_model`, which is the log-variance of the likelihood. If the latter is a constant model, then you would just get a constant.

**Reply by PaulScemama** [reply]
I'm doing classification, but the question was just out of curiosity. Very cool! Thanks for such clarification. Feel free to close
